models/sharky/solana/sharky_solana_events.sql

```diff
@@ -0,0 +1,91 @@
+{{ config(
+    alias = 'events',
+    partition_by = ['block_date'],
+    materialized = 'incremental',
+    file_format = 'delta',
+    incremental_strategy = 'merge',
+    unique_key = ['block_date', 'unique_trade_id'],
+    post_hook='{{ expose_spells(\'["solana"]\',
+                                    "project",
+                                    "sharky",
+                                    \'["ennnas"]\') }}'
+    )
+}}
+
+{%- set project_start_date = '2022-11-26' %} -- TODO change back to '2022-04-14'
+{%- set sharky_smart_contract = 'SHARKobtfF1bHhxD2eqftjHBdVSCbKo9JtgK71FhELP' %}
+
+WITH sharky_txs AS (
+        SELECT
+            tx_id AS id,
+            block_time
+        FROM {{ source('solana','account_activity') }}
+        WHERE
+            tx_success
+            {% if not is_incremental() %}
+            AND block_time >= '{{ project_start_date }}'
+            {% endif %}
+            {% if is_incremental() %}
+            -- this filter will only be applied on an incremental run
+            AND block_time >= date_trunc("day", now() - interval '1 week')
+            {% endif %}
+            AND address = '{{sharky_smart_contract}}'
+    ), events AS (
+        SELECT
+            'solana' as blockchain,
+            'sharky' as project,
+            signatures[0] as tx_hash,
+            block_date,
+            block_time,
+            CAST(block_slot AS BIGINT) as block_number,
+            (abs(post_balances[0] - pre_balances[0]) / 1e9) * p.price AS amount_usd,
+            (abs(post_balances[0] - pre_balances[0]) / 1e9) AS amount_original,
+            CAST(abs(post_balances[0] - pre_balances[0]) AS DECIMAL(38,0)) AS amount_raw,
+            filter(
+                instructions,
+                x -> x.executing_account = '{{sharky_smart_contract}}'
+              ) AS sharky_instructions,
+            CASE
+              WHEN array_contains( log_messages, 'Program log: Instruction: OfferLoan') THEN 'Offer'
+              WHEN array_contains( log_messages, 'Program log: Instruction: TakeLoan') THEN 'Take'
+              WHEN array_contains( log_messages, 'Program log: Instruction: RescindLoan') THEN 'Rescind'
+              WHEN
+                  (
+                      array_contains(log_messages, 'Program log: Instruction: RepayLoan')
+                      OR array_contains(log_messages, 'Program log: Instruction: RepayLoanEscrow')
+                  ) THEN 'Repay'
+              WHEN
+                  (
+                    array_contains(log_messages, 'Program log: Instruction: ForecloseLoan')
+                    OR array_contains(log_messages, 'Program log: Instruction: ForecloseLoanEscrow')
+                  ) THEN 'Foreclose'
+            ELSE 'Other' END as evt_type,
+            signer as user,
+            signatures[0] || '-' || id as unique_trade_id
```

Comment:
I am going to use [`tx_hash`, `id`] as unique keys, and remove `unique_trade_id`(string concatenation can be time consuming).